libr/bin/p/bin_nso.c

```diff
@@ -0,0 +1,269 @@
+/* radare2 - LGPL - Copyright 2017 - rkx1209 */
+
+#include <r_types.h>
+#include <r_util.h>
+#include <r_lib.h>
+#include <r_bin.h>
+#include <r_io.h>
+#include <r_cons.h>
+#include "../../../shlr/lz4/lz4.h"
+
+#define NSO_OFF(x) r_offsetof (NSOHeader, x)
+
+// starting at 0
+typedef struct {
+	ut32 magic;	// NSO0
+	ut32 pad0;	// 4
+	ut32 pad1;	// 8
+	ut32 pad2;	// 12
+	ut32 text_memoffset;	// 16
+	ut32 text_loc;	// 20
+	ut32 text_size;	// 24
+	ut32 pad3;	// 28
+	ut32 ro_memoffset;	// 32
+	ut32 ro_loc;	// 36
+	ut32 ro_size;	// 40
+	ut32 pad4;	// 44
+	ut32 data_memoffset;	// 48
+	ut32 data_loc;	// 52
+	ut32 data_size;	// 56
+	ut32 bss_size;	// 60
+} NSOHeader;
+
+static ut32 readLE32(RBuffer *buf, int off) {
+	int left = 0;
+	const ut8 *data = r_buf_get_at (buf, off, &left);
+	return left > 3? r_read_le32 (data): 0;
+}
+
+static uint32_t decompress(const ut8 *cbuf, ut8 *obuf, uint32_t csize, uint32_t usize) {
```

Comment:
size may result in a pretty huge number if the roff - toff is negative. better use signed sizes for this reason. otherwise you will have to check all those values on each call to this function to ensure you are not wrapping a negative number into a huge unsigned one